### DISCUSSION TITLE
Article .NET Aspire Azure Cosmos DB component: map has no entry for key "SERVICE_BINDING_CDBKV_ENDPOINT"

### DISCUSSION THREAD

**Discussion by Jeroen-VdB** [microsoft/aspire#4074] [Q&A] [answered]
I was following article [.NET Aspire Azure Cosmos DB component](https://learn.microsoft.com/en-us/dotnet/aspire/database/azure-cosmos-db-component?tabs=dotnet-cli) with a new .NET Aspire Starter Application template (Preview 6) in Visual Studio.

When deploying the application I get the following error:
> Error failed deploying service 'apiservice': failing invoking action 'deploy', failed executing template file: template: containerApp.tmpl.yaml:21:29: executing "containerApp.tmpl.yaml" at <.Env.SERVICE_BINDING_CDBKV_ENDPOINT>: map has no entry for key "SERVICE_BINDING_CDBKV_ENDPOINT"

I've been stuck for a day and I can't figure out what this issue exactly is or how to solve it. 
Does anybody have any idea?

My AppHost Program.cs:
```csharp
var builder = DistributedApplication.CreateBuilder(args);

var cosmosdb = builder.ExecutionContext.IsPublishMode
    ? builder.AddAzureCosmosDB("cdb")
    : builder.AddConnectionString("cdb");

var apiService = builder.AddProject<Projects.AspireAppWithCosmos_ApiService>("apiservice")
    .WithReference(cosmosdb);

builder.AddProject<Projects.AspireAppWithCosmos_Web>("webfrontend")
    .WithExternalHttpEndpoints()
    .WithReference(apiService);

builder.Build().Run();

```

and the ApiService Program.cs:
```csharp
var builder = WebApplication.CreateBuilder(args);

// Add service defaults & Aspire components.
builder.AddServiceDefaults(); 
builder.AddAzureCosmosDBClient("cosmosdb");

...
```

Full logs:
```
03/05/2024 17:08:49: Info 
03/05/2024 17:08:49: Info Provisioning Azure resources (azd provision)
03/05/2024 17:08:49: Info Provisioning Azure resources can take some time.
03/05/2024 17:08:49: Info 
03/05/2024 17:08:50: Info Analyzing Aspire Application (this might take a moment...)
03/05/2024 17:08:54: Info Initialize bicep provider
03/05/2024 17:08:57: Info Reading subscription and location from environment...
03/05/2024 17:08:58: Info Subscription: XXXX
03/05/2024 17:08:58: Info Location: France Central
03/05/2024 17:08:58: Info 
03/05/2024 17:08:58: Info Creating a deployment plan
03/05/2024 17:08:58: Info Comparing deployment state
03/05/2024 17:08:59: Info   (-) Skipped: Didn't find new changes.
03/05/2024 17:08:59: Info 
03/05/2024 17:08:59: Info Deploying services (azd deploy)
03/05/2024 17:08:59: Info 
03/05/2024 17:08:59: Info Deploying service apiservice
03/05/2024 17:08:59: Info Deploying service apiservice (Logging in to registry)
03/05/2024 17:09:00: Info Deploying service apiservice (Pushing container image)
03/05/2024 17:10:34: Info Deploying service apiservice (Updating container app)
03/05/2024 17:10:34: Info   (x) Failed: Deploying service apiservice
03/05/2024 17:10:34: Error failed deploying service 'apiservice': failing invoking action 'deploy', failed executing template file: template: containerApp.tmpl.yaml:21:29: executing "containerApp.tmpl.yaml" at <.Env.SERVICE_BINDING_CDBKV_ENDPOINT>: map has no entry for key "SERVICE_BINDING_CDBKV_ENDPOINT"
```

**Comment by davidfowl** [comment]
cc @vhvb1989 @ellismg

**Comment by vhvb1989** [comment, marked as answer]
Hi @Jeroen-VdB , this issue is fixed on azd daily build: https://github.com/Azure/azure-dev/blob/main/cli/installer/README.md#download-from-daily-builds.  Can you please give a try?

**Reply by Jeroen-VdB** [reply]
Using `azd version 1.9.0-beta.1-daily.3760115 (commit 31fe15a3c6b78eab820b5a515961abbcb5b18fac)` solved it.
Thank you!
Do you have any insight into what the issue might have been?

**Reply by vhvb1989** [reply]
yeah, It was a regression in azd for naming key vault.  Fixed with: https://github.com/Azure/azure-dev/pull/3822

The name strategy for kv was updated, but we missed the linked reference to the yaml deployment files. This impacted folks using Cosmos, Redis and/or SQLServer as Azure Services
